src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs

```diff
@@ -80,46 +85,49 @@ public void UpdatePool(IAccountStateProvider accounts, Func<Address, AccountStru
                 Debug.Assert(bucket.Count > 0);
 
                 accounts.TryGetAccount(address, out AccountStruct account);
-                UpdateGroupNonLocked(address, account, bucket, changingElements);
+                UpdateGroupNonLocked(address, account, bucket, updateElements);
             }
         }
 
-        private void UpdateGroupNonLocked(Address groupKey, AccountStruct groupValue, EnhancedSortedSet<Transaction> bucket, Func<Address, AccountStruct, EnhancedSortedSet<Transaction>, IEnumerable<(Transaction Tx, UInt256? changedGasBottleneck)>> changingElements)
+        private void UpdateGroupNonLocked(Address groupKey, AccountStruct groupValue, EnhancedSortedSet<Transaction> bucket, UpdateGroupDelegate updateElements)
         {
             _transactionsToRemove.Clear();
             Transaction? lastElement = bucket.Max;
 
-            foreach ((Transaction tx, UInt256? changedGasBottleneck) in changingElements(groupKey, groupValue, bucket))
+            updateElements(groupValue, bucket, ref lastElement, _updateTx);
+
+            ReadOnlySpan<Transaction> txs = CollectionsMarshal.AsSpan(_transactionsToRemove);
+            for (int i = 0; i < txs.Length; i++)
             {
-                if (changedGasBottleneck is null)
-                {
-                    _transactionsToRemove.Add(tx);
-                }
-                else if (Equals(lastElement, tx))
-                {
-                    bool reAdd = _worstSortedValues.Remove(tx);
-                    tx.GasBottleneck = changedGasBottleneck;
-                    if (reAdd)
-                    {
-                        _worstSortedValues.Add(tx, tx.Hash!);
-                    }
+                TryRemoveNonLocked(txs[i].Hash!, evicted: false, out _, out _);
+            }
+        }
 
-                    UpdateWorstValue();
-                }
-                else
+        private void UpdateTransaction(EnhancedSortedSet<Transaction> bucket, Transaction tx, in UInt256? changedGasBottleneck, ref Transaction? lastElement)
+        {
+            if (changedGasBottleneck is null)
+            {
+                _transactionsToRemove.Add(tx);
+            }
+            else if (Equals(lastElement, tx))
+            {
+                bool reAdd = _worstSortedValues.Remove(tx);
+                tx.GasBottleneck = changedGasBottleneck;
+                if (reAdd)
                 {
-                    tx.GasBottleneck = changedGasBottleneck;
+                    _worstSortedValues.Add(tx, tx.Hash!);
                 }
-            }
 
-            ReadOnlySpan<Transaction> txs = CollectionsMarshal.AsSpan(_transactionsToRemove);
-            for (int i = 0; i < txs.Length; i++)
+                UpdateWorstValue();
+                lastElement = bucket.Max;
```

Comment:
LastElement was previously never changed for comparision even though the worse value was changed

Reply:
Transactions in buckets are sorted by nonces, so changing gas bottleneck cannot change order

Reply:
To use `GetEnumerator` and have structs is fiendisly complicated as there are 3 types of Enumerators two at same level and one which needs to be embedded in the others; having them as IEnumeratable causes multiple allocations on every tx add to the pool (which is the whole tx pool every block + others)